### ISSUE TITLE
Diff recursion explosion with nested tags

### ISSUE THREAD

**Issue by alehander42** [pragmagic/karax#40]
I have a problem in a larger karax application, which was hard to reproduce, but I think this artificial example should be sufficient:

```nim

import vdom, karax, karaxdsl

type
  User = object
    name: cstring

var users = @[@[User(name: cstring"a")], @[User(name: cstring"b")]]

const MAX = 250

for z in 0..<MAX:
  users.add(@[])
  for y in 0..<20:
    users[^1].add(User(name: cstring("c" & $z)))

var specialLine = -1

proc render: VNode =
  result = buildHtml(tdiv):
    for i, line in users:
      var special = ""
      if specialLine == i:
        special = "focus"

      tdiv(class="line " & special):
        for m in line:
          span:
            text(m.name)

proc ok: VNode =
  # render()
  result = buildHtml(tdiv):
    tdiv:
      tdiv:
        render()

proc top: VNode =
  result = buildHtml(tdiv):
    section(id="main"):
      ok()

setRenderer(top)
```
I start this with `karun --run example.nim` and stats enabled and the only modifications I have are that instead of `if kxi.recursion > 100:stuff` I have `if kxi.recursion mod 1000 == 0:kout kxi.recursion`,to more easily monitor the value itself

When I call `redraw`, normally it logs ~5000 kxi.recursion which is almost instant.

When I change specialLine in the console, (e.g. to 2), the first redraw after that logs  ~1200000 kxi.recursion.

If we test with changing ok: if you just have `render()` there, you log ~40000 kxi.recursion, but for each nested tdiv in which render VNode goes, it logs 3 * more recursion.

As a whole, even the 5000 value is weird, because there is `if kxi.recursion > 100:` debug log, so I assume that may be karax is not supposed to go even to that number: I guess it has also something to do with the huge number of tags (but 250 * ~20 shoudn't be that much)

**Comment by alehander42** [comment]
Is this happening because of wrong usage of Karax? ( a.k.a. is the solution "use karax in a different intended way and you won't run into this)

**Comment by Araq** [comment]
I'm investigating right now. :-) Looks fine to me and should just work.

**Comment by Araq** [comment]
I changed it to the following program, so that mouse clicks trigger a redraw:

```nim
import vdom, karax, karaxdsl, vstyles

type
  User = object
    name: cstring

var users = @[@[User(name: cstring"a")], @[User(name: cstring"b")]]

const MAX = 250

for z in 0..<MAX:
  users.add(@[])
  for y in 0..<20:
    users[^1].add(User(name: cstring("c" & $z)))

var specialLine = -1

proc select(i: int): proc () =
  let j = i
  result = proc () =
    specialLine = j

proc render: VNode =
  result = buildHtml(tdiv):
    for i, line in users:
      let st = if specialLine == i: style(StyleAttr.color, cstring"green") else: style()

      tdiv(class="line ", onclick = select(i)):
        for m in line:
          span(style = st):
            text(m.name)

proc ok: VNode =
  # render()
  result = buildHtml(tdiv):
    tdiv:
      tdiv:
        render()

proc top: VNode =
  result = buildHtml(tdiv):
    section(id="main"):
      ok()

setRenderer(top)

```

The responsiveness remains excellent, I cannot reproduce any issues.

**Comment by alehander42** [comment]
So I can easily reproduce the issue by changing only two lines in your example:

```diff
26,28c26,27
<       let st = if specialLine == i: style(StyleAttr.color, cstring"green") else: style()
< 
<       tdiv(class="line ", style=st, onclick = select(i)):
---
>       let st = if specialLine == i: "focus" else: ""
>       tdiv(class="line " & st, onclick = select(i)):
```

Basically iff the class changes, it all goes into a lot of recursion. So is there a better way to add conditional classes ?

**Comment by alehander42** [comment]
(You wouln't be able to visually see the line changing without having `.focus { color: green; }` somewhere, but with some kind of enabled karax debugging logging, it should be easy to see at least the speed)

**Comment by Araq** [comment]
> So I can easily reproduce the issue by changing only two lines in your example

Well I started with your version. But let me try again.

>  but with some kind of enabled karax debugging logging, it should be easy to see at least the speed)

The logging itself causes a massive slowdown, watch out.

**Comment by alehander42** [comment]
Ok, I disable all logging, then if tweak the value of MAX, you can still notice the speed difference: e.g. for 1200, your version is instant and mine takes 4 seconds

**Comment by Araq** [comment]
Should be fixed.
